### ISSUE TITLE
Build Error On Eclipse [4410]

### ISSUE THREAD

**Issue by drorbashan** [eProsima/IDL-Parser#8]
Hi,

I build the IDL Parser within Eclipse Oxygen.1a Release (4.7.1a), and I am getting the following error:

missing arguments(s) on rule reference: export (org.antlr:antlr4-maven-plugin:4.2.2:antlr4:antlr:generate-sources)	
pom.xml	/IDL-Parser-master	line 355	Maven Build Participant Problem

I have loaded an older version of the IDL Parser (2015) and everything is working fine.

Any idea for what can cause the problem with the current version.

I saw something similar that was opened on Sep 2018, are there any updates?

Tx,
Dror.

**Comment by richiware** [comment]
Are you trying to build KIARA grammar parser or IDL grammar parser?

**Comment by drorbashan** [comment]
I am trying to build IDL grammar parser.
Tx, Dror.

**Comment by richiware** [comment]
`pom.xml` is outdated. Currently we use `gradle` to build the IDL grammar parser. Right now I don't have enough knowledge and the environment to update it.

**Comment by drorbashan** [comment]
Can you provide me some directions what to look for, so I can try find a solution?
Are you considering to look into it sometime soon, or maven will not be supported?
In regard to environment it is just to load the project to eclipse and you get the error.
Tx,
Dror.

**Comment by richiware** [comment]
We were talking and we will try to update the maven support next week.

**Comment by drorbashan** [comment]
Great, Tx, really appreciate it.
Dror.

**Comment by LuisGP** [comment]
Hi Dror!

I just created the PR #9 that should solve the issue with maven.
Could you test it?

Thanks!
Regards.

**Comment by drorbashan** [comment]
Hi Luis,

Thanks for the efforts, I did a quick test and here are some comments. The test includes parsing sample IDL file and going over your data model to extract basic data.
Bottom line I succeed to work with it with minor adjustment.

1.     The files: IDLBaseListener, IDLLexer, IDLListener, IDLParser are created under default package and not under com.eprosima.idl.parser.grammer, so the import for these files fails. I created new package with the right name under target/generated-source…  folder and copied the files to there and everything is fine.
2.	Annotation which was available on the version I used before is not available here so I commented the import and any related code.
3.	I upgraded the pom.xml to antlr 4.7 since I used CharStreams for generating the input for the IDLLexer. 
4.	I am attaching the file I used to do the test, if you can take a look to see if this is the right way of using the package. My main goal is to parse an IDL file with your parser and then go over your data model to fill my data model.
In the following two weeks I will integrate it more deeply into our product, will update you for any issues I will have or not have.
[CuseIDLParser.txt](https://github.com/eProsima/IDL-Parser/files/2792929/CuseIDLParser.txt)

Let me know if something is not clear.

Thanks again and Regards,
Dror.

**Comment by LuisGP** [comment]
Hi Dror,

I'm glad to see you can continue with your integration of IDL-Parser.
About your comments:

1. We'll create a task to investigate and solve that issue.

1. We support some annotations but not all of them. It is planned to implement OMG IDL 4.2, so we'll implement all the annotations that version includes.

1. Yes, we saw you used CharStream instead the deprecated ANTLRFileStream. We at this moment doesn't have planned to migrate to newer versions of antlr.

1. As I commented you through email, currently the best way to investigate how to use IDL-Parser is taking a look to projects that use it already, as *fastrtpsgen* or [Micro-XRCE-DDS-Gen](https://github.com/eProsima/Micro-XRCE-DDS-Gen). In *fastrtpsgen*, we use the TemplateManager and TemplateGroup filled by calling `IDLParser.specification` instead of using the *specification* to iterate through the definitions, but I guess you use that way because it matches best your needs (at least for testing?). I'm sad we don't have any documentation or examples for IDL-Parser and I'm afraid that create it is not planned.

We hope you achieve what you need for your product and we hope to hear good news from you soon :-).

Best regards.

**Comment by LuisGP** [comment]
Hi again, Dror

We have created a new PR #10, that should fix the issue with the package of the generated files. In addition, it adds the generation of a standalone jar for including IDL-Parser easily.

**Comment by drorbashan** [comment]
Luis,

Tried it but I don't see any change, still the files are created under the target/generated-sources/antlr4 folder and the package is (default package) .

I am using eclipse Oxygen.1a Release (4.7.1a) on Windows 10.

The support for the jar is great, probably will use it later on.

Dror

**Comment by LuisGP** [comment]
Hi Dror,

I did a quick test in Windows and Eclipse and seems that in that environment maven or antlr4 aren't capable to generate the folder structure for the generated files if the g4 files aren't in a directory structure like the package. It's weird because on Linux and "raw" maven or antlr4 creates the structure for the generated sources for us.

In any case, I uploaded the changes to a new branch until I can test it properly that it doesn't have any side effect on Linux. Can you test it, please?

Branch: [hotfix/g4_files_to_package](https://github.com/eProsima/IDL-Parser/tree/hotfix/g4_files_to_package)

Best regards.

**Comment by drorbashan** [comment]
Luis,

I did a quick test and it seems that the fix is doing the job, files are generated in the right place and full build is completed.

Tx,
Dror.

**Comment by drorbashan** [comment]
Hi Luis,

I continue integrating eProsima parser and for the moment everything is fine.
I am trying now to work with idl files that include other idl files something like:

file foo.idl
module foo { 
  struct FooStruct { 
    long id; 
  }; 
};

and then file test.idl that includes foo.idl
#include <foo.idl>
 module Test { 
    struct TestStruct {
       string a;
       FooStruct foo_field;   //  I tried also foo:: FooStruct  foo_field;
    };
};

I am getting the following message:  “error: 'FooStruct' was not defined previously”

Is idl include functionality supported? am I doing something wrong? I am using the same main as was attached before.

Second question, how do I get the list of included files? meaning how do I find that module Test has include declaration of model foo?

Class Specification and the Definition interface supports:
isIsModule();
isIsInterface();
isIsException();
isIsTypeDeclaration();
isIsConstDeclaration();
isIsAnnotation();

is the Include syntax covered by one of them?

Tx,
Dror.

**Comment by drorbashan** [comment]
Luis,

Of course I tried also with #include "foo.idl"

**Comment by LuisGP** [comment]
Hi Dror,

Both formats (#include "file.idl" and #include <file.idl>) should work.

Make sure you provide the `include paths` to the Context during its creation to allow it to find the included files. 

For example, in FastRTPSGen you can find:
`Context ctx = new Context(onlyFileName, idlFilename, **m_includePaths**, m_subscribercode, m_publishercode, m_localAppProduct, m_type_object_files);`

Take a look at the `String callPreprocessor(String idlFilename)` method too as the preprocessor is in charge to perform the inclusions.

**Comment by drorbashan** [comment]
Luis,

In regard to the include path I am doing:
ArrayList<String> includePath = new ArrayList<String>();
includePath.add("-I e:\\data\\idl");      // tried also ("-Ie:\\data\\idl")		
Context context = new Context(fileName, fullFileName, includePath);
I added print and saw that "e:\data\idl\" was added to m_includePaths.
Is this OK?

I am interested only to parse idl file , not producing any source code, so do I need any of m_subscribercode, m_publishercode, m_localAppProduct, m_type_object_files?

I am looking at the at the processPreprocessorLine method in the com.eprosima.idl.Context class.  In the comments is says “This function is call when a preprocessor line was found by the lexer.” and it actually does happen.

The method definition is:
public void processPreprocessorLine(String line, int nline)
{
    	System.out.println("processPreprocessorLine " + line + " " + nline); // added by me
        // If there is a line referring to the content of an included file.
         if(line.startsWith("# ")) 
        {
            String line_ = line.substring(2);

            /* The received preprocessor line has the following form:
             * ' numline filename flags'
             * where:
             * - numline Number of the line where the include was.
             * - filename The filename whose content was included.
             * - flags
             */
I added the system print to see what it gets and it is:
#include "foo.idl"
 2

The if statement after that looks for something that starts with “# “  (with a space) which does not match the input it gets so the line is ignored, also the comments after that describe a format that doesn’t match what it receives, so I was thinking maybe I am missing some configuration? Or the #include line is parsed in another place?

Any suggestions?

**Comment by drorbashan** [comment]
Luis,
I am attaching TestIDLParser.java and two idl files to reconstruct the issue I am getting.
I really appreciate if you can take a look and help to solve it.
Running environment is eclipse on Windows 10.
Tx,
Dror.
[testfiles.zip](https://github.com/eProsima/IDL-Parser/files/2830988/testfiles.zip)

**Comment by drorbashan** [comment]
Luis,
Please let me know if someone will take a look at it.
Tx,
Dror.

**Comment by LuisGP** [comment]
Hi Dror,
I was sick the last two days. I'll try to take a look at your problem soon.

Best regards.

**Comment by LuisGP** [comment]
Hi Dror,

I took a quick look and seems that FastRTPSGen calls the preprocessor explicitly before creating the Context. It should be easy to adapt the method to adjust your needs.

So your code should look something like:

```java
		
    String idlParseFileName = callPreprocessor(fullFileName, includePath, null);
        
    if (idlParseFileName == null)
    {
        System.out.println("Error calling preprocessor.");
        return;
    }
		
    Context context = new Context(fileName, fullFileName, includePath);
				
    try {
			
        ANTLRFileStream input = new ANTLRFileStream(idlParseFileName);    

```

This `callPreprocessor` method tries to call "cl.exe" on Windows, so make sure to provide your program an environment where it is reacheable, like a VS terminal.

Hope this helps you to keep going.
Best regards.

**Comment by drorbashan** [comment]
Tx,
I will try and update, but why do I need the cl.exe, is it use to generate some files, I am not going to create or compile any source files, just parse the idl file?

**Comment by LuisGP** [comment]
Seems that cl.exe is used just to solve the preprocessor directives.

**Comment by drorbashan** [comment]
Hi Luis,
I followed your guidance and made it work, thanks a lot for the support and efforts.
I am attaching the final main I used for testing, if someone would like to only parse idl files he may use it.
Again, Tx a lot.
Dror.
[TestIDLParser.zip](https://github.com/eProsima/IDL-Parser/files/2851388/TestIDLParser.zip)

**Comment by drorbashan** [comment]
Hi Luis,

You have mentioned that annotations are currently not supported, any plans for adding it? 
Comments that exists in idl files, I noticed they exist in the .cc files created as part of the process but I didn’t find how to get them from the Java data model, is it possible?

Tx,
Dror.

**Comment by LuisGP** [comment]
Hi Dror,

yes, we are working in IDL 4.2 implementation, including annotations support.

IDL Comments are sent to antlr4 "HIDDEN" channel. These .cc files are the output of the preprocessor, so IDL-Parser doesn't have any control over them.

If you want to capture the comments, you should add a custom rule in the IDL.g4 file, COMMENT, and LINE_COMMENT rules, or access to that "HIDDEN" channel, if possible.

**Comment by LuisGP** [comment]
Hi Dror,

we just uploaded the annotations feature to the branch feature/IDL_4.2.
In that branch, you will find your TestIDLParser at com.eprosima.idl.test package, a little modified to our testing purposes. There, you can take a look at how annotations can be accessed from TypeDeclarators and Members. Annotations should be accessible directly through TypeCode too.

If you test it and find any issue, please report them.

Thank you very much, best regards.

**Comment by drorbashan** [comment]
Luis,
I will look and try it, just to ensure this branch includes the fixes from hotfix/g4_files_to_package?
Tx,
Dror.

**Comment by drorbashan** [comment]
Did very basic testing and it looks fine.
If you have samples of idl files with annotations you would like me to run in my environment just send them.

Tx,
Dror.

**Comment by LuisGP** [comment]
Hi Dror,
Yes, it includes these fixes (PR #10).
Currently, we are testing with some files included in "test/idls". The newer file is "forward_decls".

Thank you.
Luis GP.

**Comment by LuisGP** [comment]
I'm closing this issue due the original question is solved and a lot of improvements have been made since then :).
Feel free to open another issue to contact us with any other question or problem.
